### PR TITLE
Add padding_width to montage2d and add montage_rgb

### PR DESCRIPTION
**Pull request by sciunto**
## Description

#2179 was ready to be merged, but never merged (almost a year ago). I made the necessary changes (rebase and tests).

@scikit-image/core please edit and review

Note that arr_out is forced to have the same type as arr_in. Let's check that we all agree with that.

Closes https://github.com/scikit-image/scikit-image/issues/2632.

### CODE REVIEW

File: skimage/util/montage.py [hunk 1]
```diff
@@ -62,26 +66,36 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
     (4, 4)
     >>> arr_out
-    array([[  0. ,   1. ,   4. ,   5. ],
-           [  2. ,   3. ,   6. ,   7. ],
-           [  8. ,   9. ,   5.5,   5.5],
-           [ 10. ,  11. ,   5.5,   5.5]])
+    array([[ 0,  1,  4,  5],
+           [ 2,  3,  6,  7],
+           [ 8,  9,  5,  5],
+           [10, 11,  5,  5]])
```
Comment (soupault) [inline skimage/util/montage.py]: Filled with `5` here, but mean is still `5.5` (see below). Doesn't seem correct.

File: CONTRIBUTORS.txt [hunk 2]
```diff
@@ -250,3 +250,9 @@
 
 - Gleb Goussarov
   Chan-Vese Segmentation
+
+- Scott Sievert
```
Comment (soupault) [inline CONTRIBUTORS.txt]: Scott is already mentioned in this list.

File: skimage/util/montage.py [hunk 3]
```diff
@@ -62,26 +66,36 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
     assert arr_in.ndim == 3
 
-    n_images, height, width = arr_in.shape
+    # -- fill missing patches (needs to be calculated before border padding)
+    if fill == 'mean':
+        fill = arr_in.mean()
 
-    arr_in = arr_in.copy()
+    # -- add border padding, np.pad does all dimensions
+    # so we remove the padding from the first
```
Comment (soupault) [inline skimage/util/montage.py]: from the first ... ?

File: skimage/util/montage.py [hunk 4]
```diff
@@ -108,3 +121,68 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
     arr_out = arr_out.reshape(alpha_y * height, alpha_x * width)
 
     return arr_out
+
+
+def montage_rgb(arr_in, fill='mean', grid_shape=None, padding_width=0):
```
Comment (soupault) [inline skimage/util/montage.py]: How about naming this function `montage_multichannel`?

File: skimage/util/montage.py [hunk 5]
```diff
@@ -62,26 +66,36 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
+    if fill == 'mean':
+        fill = arr_in.mean()
 
-    arr_in = arr_in.copy()
+    # -- add border padding, np.pad does all dimensions
+    # so we remove the padding from the first
+    if padding_width > 0:
+        # only pad after to make the width correct
+        bef_aft = (0, padding_width)
+        arr_in = np.pad(arr_in, ((0, 0), bef_aft, bef_aft), mode='constant')
```
Comment (soupault) [inline skimage/util/montage.py]: I sitll think we should pad on all sides of the array (i.e. `bef_aft = (padding_width, padding_width)`)

File: skimage/util/montage.py [hunk 6]
```diff
@@ -108,3 +121,68 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
+
+def montage_rgb(arr_in, fill='mean', grid_shape=None, padding_width=0):
+    """Create a 3-dimensional 'montage' from a 4-dimensional input array
+    representing an ensemble of equally shaped 3-dimensional images.
+
+
+    Parameters
+    ----------
+    arr_in : ndarray, shape=[n_images, height, width, n_channels]
+        3-dimensional input array representing an ensemble of n_images
```
Comment (soupault) [inline skimage/util/montage.py]: 4-dimensional

File: skimage/util/montage.py [hunk 7]
```diff
@@ -108,3 +121,68 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
+        of equal shape (i.e. [height, width, n_channels]).
+    fill : float or 'mean', optional
+        How to fill the 2-dimensional output array when sqrt(n_images)
```
Comment (soupault) [inline skimage/util/montage.py]: 3-dimensional

File: skimage/util/montage.py [hunk 8]
```diff
@@ -108,3 +121,68 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
+        The default aspect ratio is square.
+    padding_width : int, optional
+        The size of the spacing between the tiles to make the
+        boundaries of individual frames easier to see.
+
+    Returns
+    -------
+    arr_out : ndarray, shape=[alpha * height, alpha * width, n_channels]
+        Output array where 'alpha' has been determined automatically to
+        fit (at least) the `n_images` in `arr_in`.
```
Comment (soupault) [inline skimage/util/montage.py]: This sounds really weird. Do we `fit in arr_in`? Don't think so.

File: skimage/util/montage.py [hunk 9]
```diff
@@ -108,3 +121,68 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
+
+
+def montage_rgb(arr_in, fill='mean', grid_shape=None, padding_width=0):
+    """Create a 3-dimensional 'montage' from a 4-dimensional input array
+    representing an ensemble of equally shaped 3-dimensional images.
+
+
+    Parameters
+    ----------
+    arr_in : ndarray, shape=[n_images, height, width, n_channels]
```
Comment (soupault) [inline skimage/util/montage.py]: Should use standard notation here: `(K, M, N[, C])`, and move the detailed description to the line below.

File: skimage/util/montage.py (hunk 4 above)
Comment (soupault) [inline skimage/util/montage.py]: On a second thought, I'd suggest to create `montage` function with `multichannel` argument, incorporate 2d and 3d (and, potentially, nD) cases into it, and deprecate `montage_2d`.

If we agree on this, I could take care of the implementation.

File: skimage/util/montage.py (hunk 1 above)
Comment (sciunto) [inline skimage/util/montage.py]: It's my remark from the first post. It's cast as a int. Was like that in the original PR, and I'm not sure at all it's the right option. arr_out has the same dtype as arr_in, and arr_in is an int.

File: skimage/util/montage.py (hunk 4 above)
Comment (sciunto) [inline skimage/util/montage.py]: Sounds a great idea, bit I'm really off these nD discussions as I never use this feature at all.

File: CONTRIBUTORS.txt (hunk 2 above)
Comment (sciunto) [inline CONTRIBUTORS.txt]: I screwed  when I fixed the merge conflict.

File: skimage/util/montage.py (hunk 4 above)
Comment (soupault) [inline skimage/util/montage.py]: @sciunto I agree that there is no clear use for nD case here, but it should be relatively easy to generalize the function.
As for `multichannel`, I see a nice use for this function in multispectral satellite imaging.

**Comment by soupault** [comment]
I was thinking what if we add a parameter to this function to montage not only the images from collection, but also the channels? Example:
```
img_collect = np.empty((4, 1, 1, 3))  # 4 units of 1x1 px RGB images
mont = montage(grid_shape=(2, 2))  # the representation below may be not correct :)
    [[[0, 0, 0]]], [[[1, 1, 1]]], 
    [[[2, 2, 2]]], [[[3, 3, 3]]]
mont = montage(grid_shape=(6, 2), disassemble_channels=True)
    [0], [0], [0], [1], [1], [1], 
    [2], [2], [2], [3], [3], [3]
```

**Comment by soupault** [comment]
Okay, here's what I've done:
- Rebased on a recent `master`;
- Moved `montage2d` from `skimage.util.montage.montage2d` to `skimage.util.montage2d` (renamed `montage.py` to `_montage.py` in the process);
- Exposed `montage2d` in the corresponding `skimage/util/__init__.py`;
- Deprecated `montage2d` in favour of `montage` (new function, see below) only for 1 release (i.e. for removal in 0.15);

- Removed the implementation of `montage_rgb` and wrote the new, more clean montage implementation, which is now accepting `multichannel` argument, performs padding between images and on the all 4 borders, uses memory more efficiently. Also, there were some bugs in `montage2d` which are fixed in `montage`;
- Significantly improved the docstring;
- Refactored the tests for `montageXX` functions, and added __a lot__ of new tests for `montage`;
- Updated the gallery examples to use `montage` instead of `montage2d`.

- Fixed some imports (`color/colorlabel` and `exposure/_adapthist`) to avoid cyclic dependencies with the new availability of `montageXX` functions (i.e. as `skimage.util.montage2d`, `skimage.util.montage`).

_To summarize:_
This PR implements new `skimage.util.montage` function which operates in a similar way as previous `skimage.util.montage.montage2d`, but also provides `padding_width` and `multichannel` (now working or multichannel images) arguments.
I'd insist on breaking the API here: moving `skimage.util.montage.montage2d` to `skimage.util.montage2d`, and having a short deprecation cycle for `montage2d` function. There is absolutely no need to hide a single function under a separate module.

@scikit-image/core This is pretty much complete now, and ready for review.

File: skimage/color/colorlabel.py [hunk 10]
```diff
@@ -3,7 +3,7 @@
 import numpy as np
 
 from .._shared.utils import warn
-from .. import img_as_float
+from ..util.dtype import img_as_float
```
Comment (jni) [inline skimage/color/colorlabel.py]: Did I miss this move? I think `util.img_as_float` would be sufficiently nested?

File: skimage/util/_montage.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+
+    if multichannel:
+        return arr_out
+    else:
+        return arr_out[..., 0]
+
+
+@deprecated('montage', removed_version='0.15')
+def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None,
+              padding_width=0):
```
Comment (jni) [inline skimage/util/_montage.py]: Why doesn't this just call `montage`?

File: skimage/util/_montage.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+    # Rescale intensity if necessary
+    if rescale_intensity:
+        for i in range(n_images):
+            arr_in[i] = exposure.rescale_intensity(arr_in[i])
+
+    # Calculate the fill value
+    if fill == 'mean':
+        fill = arr_in.mean(axis=(0, 1, 2))
+    fill *= np.ones(1)
```
Comment (jni) [inline skimage/util/_montage.py]: What's going on here?

File: skimage/color/colorlabel.py (hunk 10 above)
Comment (soupault) [inline skimage/color/colorlabel.py]: Perhaps :).

File: skimage/util/_montage.py (hunk 11 above)
Comment (soupault) [inline skimage/util/_montage.py]: I just wanted to keep some backward compatibility. There is a slight difference between `montage` and `montage2d` in the way the padding is used - `montage2d` doesn't pad upper and left borders.

File: skimage/util/_montage.py (hunk 12 above)
Comment (soupault) [inline skimage/util/_montage.py]: If `arr_in` is grayscale and `fill` is a scalar, it is neccessary to upcast `fill` to a ndarray of ndim 1 (np.asarray(scalar) returns an array of ndim 0 and doesn't allow indexing). The indexing is used in https://github.com/scikit-image/scikit-image/pull/2626/files/95a8513cfaee0e99cf85f75c091a7575e89499c8#diff-81a8a5406b37c24a621cf34d9208fcc0R115

Comment (jni) [inline skimage/util/_montage.py]: @soupault Got it. Well, I've just been playing around with this, and I suggest that a much more readable option is `fill = np.atleast_1d(fill)`. Would you mind changing that?

Comment (jni) [inline skimage/util/_montage.py]: (Indeed I'm amazed that `<float> *= <array>` even works!)

Comment (soupault) [inline skimage/util/_montage.py]: @jni all fixed.

**Comment by soupault** [comment]
@scikit-image/core anyone else?

**Comment by soupault** [comment]
@scikit-image/core kind ping.

File: skimage/util/_montage.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+from .. import exposure
+from .._shared.utils import deprecated
+
+
+__all__ = ['montage', 'montage2d']
+
+
+def montage(arr_in, fill='mean', rescale_intensity=False, grid_shape=None,
+            padding_width=0, multichannel=False):
+    """Create a 2-dimensional 'montage' from a 3-/4-dimensional input array
```
Comment (stefanv) [inline skimage/util/_montage.py]: 2-dimensional is overloaded here; perhaps "Create a montage / rectangular collage" instead of "2-dimensional 'montage'".  The dimensionality here gets confusing when, directly after, 3-/4-dimensional is mentioned.

This also does not correspond to the single line, single sentence docstring format.

Suggested:

"Create a montage of several grey or color images."

And then a longer second paragraph.

File: skimage/util/_montage.py [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+    | 1 | 2 | 3 |   will return   +---+---+
+    +---+---+---+                 | 3 | * |
+       ---> K                     +---+---+
+
+    where the '*' patch will be determined by the `fill` parameter.
+
+    Parameters
+    ----------
+    arr_in : (K, M, N[, C]) ndarray
+        3-/4-dimensional input array representing an ensemble of `K` images
```
Comment (stefanv) [inline skimage/util/_montage.py]: The 3-/4- dimensional does not add much here; remove.

File: skimage/util/_montage.py [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+        of equal shape.
+    fill : float or array-like of floats or 'mean', optional
+        Value to fill the padding areas and/or the extra tiles in
+        the output array. Has to be `float` for single channel collections.
+        For multichannel collections has to be an array-like of shape of
+        number of channels. If `mean`, uses the mean value over all images.
+    rescale_intensity : bool, optional
+        Whether to rescale the intensity of each image to [0, 1].
+    grid_shape : tuple, optional
+        The desired grid shape for the montage (ntiles_row, ntiles_column).
```
Comment (stefanv) [inline skimage/util/_montage.py]:
```
`(ntiles_row, ntiles_column)`
```

File: skimage/util/_montage.py [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+        The size of the spacing between the tiles and between the tiles and
+        the borders. If non-zero, makes the boundaries of individual images
+        easier to perceive.
+    multichannel : boolean, optional
+        If True, the last `arr_in` dimension is threated as a color channel,
+        otherwise as spatial.
+
+    Returns
+    -------
+    arr_out : (K*M+p, K*N+p[, C]) ndarray
```
Comment (stefanv) [inline skimage/util/_montage.py]: K * (M + p) + p ?

File: skimage/util/_montage.py [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+    else:
+        arr_in = np.asarray(arr_in)[..., np.newaxis]
+
+    if arr_in.ndim != 4:
+        raise ValueError('Input array has to be either 3- or 4-dimensional')
+
+    n_images, n_rows, n_cols, n_chan = arr_in.shape
+
+    if grid_shape:
+        ntiles_row, ntiles_col = map(int, grid_shape)
```
Comment (stefanv) [inline skimage/util/_montage.py]:
```
[int(s) for s in grid_shape]
```

**Comment by soupault** [comment]
@stefanv Thanks for the review! All addressed. Let see if Sphinx is happy with the changes...

File: skimage/util/_montage.py [hunk 18]
```diff
@@ -8,7 +8,9 @@
 
 def montage(arr_in, fill='mean', rescale_intensity=False, grid_shape=None,
             padding_width=0, multichannel=False):
-    """Create a 2-dimensional 'montage' from a 3-/4-dimensional input array
+    """Create a montage of several single- or multichannel images.
+
+    Create a 2-dimensional 'montage' from a 3-/4-dimensional input array
```
Comment (stefanv) [inline skimage/util/_montage.py]: Still 2-dimensional 'montage' here

**Comment by soupault** [comment]
@stefanv I've reformulated the sentence. Please, check.

File: doc/release/release_dev.rst [hunk 19]
```diff
@@ -17,22 +17,23 @@ New Features
...
 ------------
 
 
 
-
 API Changes
 -----------
-
-
+- ``skimage.util.montage.montage2d`` is now available as ``skimage.util.montage2d``.
```
Comment (stefanv) [inline doc/release/release_dev.rst]: Typo

Comment (soupault) [inline doc/release/release_dev.rst]: Sorry, I do not see it... Any hints?

Comment (stefanv) [inline doc/release/release_dev.rst]: Because I misread ;)  I thought you meant to write `montage2d` -> `montage`.  My mistake!

Comment (soupault) [inline doc/release/release_dev.rst]: This is "good-to-go" then!

**Comment by stefanv** [comment]
:rocket:
